src/main/protobuf/bazel_output_service.proto

```diff
@@ -0,0 +1,349 @@
+  // requested. This means that this list has the same length as
+  // StageArtifactsRequest.artifacts.
+  repeated Response responses = 1;
+}
+
+message FinalizeArtifactsRequest {
+  message Artifact {
+    // path is relative to StartBuildResponse.output_path.
+    string path = 1;
+    // Expected digest for this file. This allows server to detect if the file
+    // has been changed after Bazel finished creating the file and the
+    // corresponding FinalizeArtifactsRequest is processed. This field MAY be
+    // set for regular files, and MUST be omitted for directories and symlinks.
+    //
+    // The concrete type of the locator depending on the CAS Bazel connects to.
+    // In case of a REv2 CAS, the type is
+    // [FileArtifactLocator][bazel_output_service_rev2.FileArtifactLocator].
+    google.protobuf.Any locator = 2;
+  }
+
+  // The identifier of the build. Server uses this to determine which output
+  // path needs to be modified.
+  string build_id = 1;
+
+  repeated Artifact artifacts = 2;
+}
+
+message FinalizeArtifactsResponse {
+  // Intentionally empty for future extensibility.
+}
+
+message FinalizeBuildRequest {
+  // The identifier of the build that should be finalized.
+  string build_id = 1;
+
+  // Whether the build completed successfully. The remote output service MAY,
+  // for example, use this option to apply different retention policies that
+  // take the outcome of the build into account.
+  bool build_successful = 2;
+}
+
+message FinalizeBuildResponse {
+  // Intentionally empty for future extensibility.
+}
+
+message BatchStatRequest {
+  // The identifier of the build. The remote output service uses this to
+  // determine which output path needs to be inspected.
+  string build_id = 1;
+
+  // Paths whose status needs to be obtained. The server MUST canonicalize the
+  // path using lstat semantics and StartBuildRequest.output_path_aliases, i.e.
+  // components except the last one must be resolved if they are symlinks. If
+  // the symlink pointing to a location outside of the output path, but is a
+  // StartBuildRequest.output_path_aliases, the server MAY use the alias to
+  // continue the canonicalization.
+  //
+  // Refer to Stat.type for how to handle a situation where canonicalization
+  // fails due to a symlink pointing to a location outside of the output path.
+  //
+  // Path is relative to StartBuildResponse.output_path.
+  repeated string paths = 2;
+}
+
+message BatchStatResponse {
+  message StatResponse {
+    // The status of the file. If no file exists at the requested path, this
+    // field MUST be unset.
+    Stat stat = 1;
+  }
+
+  message Stat {
+    message File {
+      // The digest of the file.
+      //
+      // The concrete type of the digest depending on the CAS Bazel connects to.
+      // In case of a REv2 CAS, the type is
+      // [Digest][build.bazel.remote.execution.v2.Digest].
+      google.protobuf.Any digest = 1;
```

Comment:
If the Remote Output Service knows a path corresponds to a file, but is unable to compute its digest, is it permitted to leave this field null? bb_clientd can run into these cases if files are still opened for writing. The page cache may still hold on to pages of data that need to be flushed. Could we document this explicitly?

Reply:
The server can return `null` in this case. Updated the comment.

Replaced the Digest with locator.